playroom/components/loader.tsx

```diff
@@ -0,0 +1,68 @@
+import * as React from 'react';
+
+type Props = {
+    load: string | (() => Promise<any>);
+    render: (data: any) => React.ReactElement;
+    renderLoading?: () => React.ReactElement;
+    renderError?: () => React.ReactElement;
+};
+
+type LoaderState = 'loading' | 'error' | 'success';
+
+const Loader: React.FC<Props> = ({load, render, renderLoading = () => null, renderError = () => null}) => {
+    const [loaderData, setLoaderData] = React.useState(null);
+    const [loaderStatus, setLoaderStatus] = React.useState<LoaderState>('loading');
+
+    React.useEffect(() => {
+        let isRenderActive = true;
+
+        if (isRenderActive) setLoaderStatus('loading');
+
+        if (typeof load === 'string') {
+            fetch(load)
```

Comment:
to avoid repeating some logic, you can create a `load` function when a string is provided, this way you only have a single code path:
```tsx
const loadFn = typeof load === 'function' ?  load : () => {
    fetch(load).then((response) => {
        if (!response.ok) {
            throw new Error('fetch failed');
        }
        return response.json();
    });
}

// from here you have a single code path instead of two (one for string and one for function)
```